src/combinatorics/simple_graph/matching.lean

```diff
@@ -0,0 +1,63 @@
+/-
+Copyright (c) 2020 Alena Gusakov. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Alena Gusakov.
+-/
+import data.fintype.basic
+import data.sym2
+import combinatorics.simple_graph.basic
+/-!
+# Matchings
+
+
+## Main definitions
+
+* a `matching` on a simple graph is a subset of its edge set such that
+   no two edges shares an endpoint.
+
+* a `perfect_matching` on a simple graph is a matching in which every
+   vertex belongs to an edge.
+
+TODO:
+  - Lemma stating that the existence of a perfect matching on `G` implies that
+    the cardinality of `V` is even (assuming it's finite)
+  - Hall's Marriage Theorem
+  - Tutte's Theorem
+
+TODO: Tutte and Hall require a definition of subgraphs.
+-/
+open finset
+universe u
+
+namespace simple_graph
+variables {V : Type u} (G : simple_graph V) (M : set (sym2 V)) (h : M ⊆ edge_set G)
+
+instance : inhabited (set (sym2 V)) :=
+  { default := (complete_graph V).edge_set }
+
+/--
+A matching on `G` is a subset of its edges such that no two edges share a vertex.
+-/
+def matching : Type* :=
+  { M : set (sym2 V) // M ⊆ edge_set G ∧ ∀ x ∈ M, ∀ y ∈ M, ∀ v : V, v ∈ x ∧ v ∈ y → x = y }
+
+instance : inhabited (matching G) := { default :=
+  ⟨(∅ : set (sym2 V)),
+    begin
+      split,
+      exact set.empty_subset G.edge_set,
+      intros x hx y hy v hv,
+      have h2 := set.mem_empty_eq x,
+      by_contra,
+      rw ← h2,
+      exact hx,
+    end ⟩ }
+
+/--
+A perfect matching `M` on graph `G` is a matching such that
+  every vertex is contained in an edge of `M`.
+-/
+def perfect_matching (M : matching G) :
+  Prop := ∀ v : V, ∃ x ∈ M.1, v ∈ x
```

Comment:
If possible it's probably a good idea to use a coercion `(M : set (sym2 V))` rather than projection `M.1` since the former is simp-normal form

Reply:
If we're doing coercions is there a purpose to having the `matching` type though?

Reply:
@agusakov That seems to be a good point:
```lean
def matching_verts (h : G.is_matching M) : set V :=
⋃₀(set.image coe M)

def is_perfect_matching (h : G.is_matching M) : Prop :=
G.matching_verts M h = set.univ
```
It's hard to say ahead of time whether it's better or worse having a `matching` type or just referring to subsets of `sym2` that satisfy `is_matching`.

Reply:
My guess is, since I'll be using it to try to prove theorems that focus a lot on matchings, having them be a type would probably make things easier

Reply:
I think I'll put a pin in this and consider it later - I'm gonna try to prove stuff and see how it works out as I go for now